error_writer.go

```diff
@@ -54,69 +52,54 @@ type ErrorWriter struct {
 // Options supplied via [WithConditionalHandlerOptions] are ignored.
 func NewErrorWriter(opts ...HandlerOption) *ErrorWriter {
 	config := newHandlerConfig("", StreamTypeUnary, opts)
-	writer := &ErrorWriter{
+	codecs := newReadOnlyCodecs(config.Codecs)
+	return &ErrorWriter{
 		bufferPool:                   config.BufferPool,
-		protobuf:                     newReadOnlyCodecs(config.Codecs).Protobuf(),
-		grpcContentTypes:             make(map[string]struct{}),
-		grpcWebContentTypes:          make(map[string]struct{}),
-		unaryConnectContentTypes:     make(map[string]struct{}),
-		streamingConnectContentTypes: make(map[string]struct{}),
+		protobuf:                     codecs.Protobuf(),
+		handleGRPC:                   config.HandleGRPC,
+		handleGRPCWeb:                config.HandleGRPCWeb,
 		requireConnectProtocolHeader: config.RequireConnectProtocolHeader,
 	}
-	for name := range config.Codecs {
-		unary := connectContentTypeFromCodecName(StreamTypeUnary, name)
-		writer.unaryConnectContentTypes[unary] = struct{}{}
-		streaming := connectContentTypeFromCodecName(StreamTypeBidi, name)
-		writer.streamingConnectContentTypes[streaming] = struct{}{}
-	}
-	if config.HandleGRPC {
-		writer.grpcContentTypes[grpcContentTypeDefault] = struct{}{}
-		for name := range config.Codecs {
-			ct := grpcContentTypeFromCodecName(false /* web */, name)
-			writer.grpcContentTypes[ct] = struct{}{}
-		}
-	}
-	if config.HandleGRPCWeb {
-		writer.grpcWebContentTypes[grpcWebContentTypeDefault] = struct{}{}
-		for name := range config.Codecs {
-			ct := grpcContentTypeFromCodecName(true /* web */, name)
-			writer.grpcWebContentTypes[ct] = struct{}{}
-		}
-	}
-	return writer
 }
 
 func (w *ErrorWriter) classifyRequest(request *http.Request) protocolType {
 	ctype := canonicalizeContentType(getHeaderCanonical(request.Header, headerContentType))
-	if _, ok := w.unaryConnectContentTypes[ctype]; ok {
-		if err := connectCheckProtocolVersion(request, w.requireConnectProtocolHeader); err != nil {
-			return unknownProtocol
+	method := request.Method
+	switch {
+	case w.handleGRPC && (ctype == grpcContentTypeDefault || strings.HasPrefix(ctype, grpcContentTypePrefix)):
+		if method != http.MethodPost {
+			break
+		}
+		return grpcProtocol
+	case w.handleGRPCWeb && (ctype == grpcWebContentTypeDefault || strings.HasPrefix(ctype, grpcWebContentTypePrefix)):
+		if method != http.MethodPost {
+			break
+		}
+		return grpcWebProtocol
+	case strings.HasPrefix(ctype, connectStreamingContentTypePrefix):
+		if method != http.MethodPost {
+			break
 		}
-		return connectUnaryProtocol
-	}
-	if _, ok := w.streamingConnectContentTypes[ctype]; ok {
 		if err := connectCheckProtocolVersion(request, w.requireConnectProtocolHeader); err != nil {
-			return unknownProtocol
+			break
```

Comment:
Just want to flag that there's an open discussion about this in #700. Let's make sure we sync this up with whatever's decided in that thread before merging to `main`.